oslo/transformers/trainer.py

```diff
@@ -0,0 +1,852 @@
+        inputs = self._prepare_inputs(inputs)  # TODO Check
+        # log_dist(f"After self._prepare_inputs: \n{inputs}", rank=-1)
+
+        with self.autocast_smart_context_manager():
+            loss = self.compute_loss(model, inputs)
+
+        if self.args.n_gpu > 1:
+            loss = loss.mean()  # mean() to average on multi-gpu parallel training
+
+        if self.args.gradient_accumulation_steps > 1:
+            # deepspeed handles loss scaling by gradient_accumulation_steps in its `backward`
+            loss = loss / self.args.gradient_accumulation_steps
+
+        if self.do_grad_scaling:
+            self.scaler.scale(loss).backward()
+        else:
+            loss.backward()
+
+        return loss.detach()
+
+    def _wrap_model(self, model_wrappers: List, training: bool = True):
+        if not training:
+            return self.model
+        if unwrap_model(self.model) is not self.model:
+            return self.model
+
+        model = self.model
+        # Distributed training (should be after apex fp16 initialization)
+        if self.parallel_context is not None:
+            for wrapper in model_wrappers:
+                log_dist(f"Model wrapping with wrapper: {wrapper}")
+
+                if wrapper == TensorParallel:
+                    log_dist(self.args.oslo_config)
+                    model = wrapper(
+                        model,
+                        parallel_context=self.parallel_context,
+                        **self.args.oslo_config.tensor_parallelism["params"],
+                    )
+                elif wrapper == PipelineParallel:
+                    model = wrapper(
+                        model,
+                        parallel_context=self.parallel_context,
+                        **self.args.oslo_config.pipeline_parallelism["params"],
+                    )
+                elif wrapper == SequenceParallel:
+                    model = wrapper(
+                        model,
+                        parallel_context=self.parallel_context,
+                        **self.args.oslo_config.sequence_parallelism["params"],
+                    )
+                elif wrapper == DistributedDataParallel:
+                    # model = model.to()
+                    model = wrapper(
+                        model,
+                        parallel_context=self.parallel_context,
+                        **self.args.oslo_config.data_parallelism["params"],
+                    )
+                elif wrapper == DataParallel:
+                    self.create_optimizer()
+                    model, self.optimizer = wrapper(
+                        model,
+                        self.optimizer,
+                        parallel_context=self.parallel_context,
+                        zero_stage=self.args.oslo_config["data_parallelism"]["zero_stage"],
+                    )
+                log_dist(f"Model wrapping with {wrapper}")
+
+            allocate_params(model, self.parallel_context)
```

Comment:
please use `oslo.ready(model, parallel_context)` !

Reply:
it is replaced